### PR TITLE
Compile time sanity check to avoid runtime error.

### PR DESCRIPTION
**Pull request by srkreddy1238**
Check output dimensions also match
This check is to make sure the NNVM operator Infer match with Compute result.
This helps to avoid developers misguided with runtime errors instead of compile time.

### CODE REVIEW

File: src/compiler/compile_engine.cc [hunk 1]
```diff
@@ -202,6 +202,24 @@ class CompileEngine {
       Array<Tensor> out = fcompute[inode.source->op()](
           inode.source->attrs, op_inputs, out_info);
       CHECK_EQ(out.size(), inode.source->num_outputs());
+
+      // check output dimentions also match
+      // This check is to make sure the NNVM operator Infer match with Compute result.
+      // Missing this check may pass the build but leads to runtime errors.
+      for (uint32_t i=0; i < out.size(); i++) {
```
Comment (tqchen) [inline src/compiler/compile_engine.cc]: prefer ++i over i++

File: src/compiler/compile_engine.cc [hunk 2]
```diff
@@ -202,6 +202,24 @@ class CompileEngine {
...
+        CHECK_EQ(out[i].ndim(), out_info[i].ndim()) << inode.source->op()->name;
+        tvm::Tensor in_tensor = out[i];
```
Comment (tqchen) [inline src/compiler/compile_engine.cc]: the name in_tensor is confusing here

File: src/compiler/compile_engine.cc (hunk 1 above)
Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: done

File: src/compiler/compile_engine.cc (hunk 2 above)
Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: Updated with inferred_tensor, computed_tensor instead :)

File: src/compiler/compile_engine.cc [hunk 3]
```diff
@@ -202,6 +202,24 @@ class CompileEngine {
...
       CHECK_EQ(out.size(), inode.source->num_outputs());
+
+      // check output dimentions also match
+      // This check is to make sure the NNVM operator Infer match with Compute result.
+      // Missing this check may pass the build but leads to runtime errors.
+      for (uint32_t i=0; i < out.size(); ++i) {
+        CHECK_EQ(out[i].ndim(), out_info[i].ndim()) << inode.source->op()->name;
+        tvm::Tensor inferred_tensor = out[i];
+        tvm::Tensor computed_tensor = out_info[i];
+        // TODO(srkreddy1238): To be verified before enabling it.
```
Comment (tqchen) [inline src/compiler/compile_engine.cc]: Remove the following lines, output shape cannot equal input shape

File: src/compiler/compile_engine.cc [hunk 4]
```diff
@@ -202,6 +202,24 @@ class CompileEngine {
...
+        // Especially for operator like cast with different dtype for input and output.
+        // CHECK_EQ(inferred_tensor->dtype, computed_tensor->dtype);
+        for (uint32_t j = 0; j < inferred_tensor->shape.size(); j++) {
+        if ((as_const_int(inferred_tensor->shape[j])) && (as_const_int(computed_tensor->shape[j])))
```
Comment (tqchen) [inline src/compiler/compile_engine.cc]: alignment issue

File: src/compiler/compile_engine.cc [hunk 5]
```diff
@@ -202,6 +202,24 @@ class CompileEngine {
...
+      // This check is to make sure the NNVM operator Infer match with Compute result.
+      // Missing this check may pass the build but leads to runtime errors.
+      for (uint32_t i=0; i < out.size(); ++i) {
+        CHECK_EQ(out[i].ndim(), out_info[i].ndim()) << inode.source->op()->name;
+        tvm::Tensor inferred_tensor = out[i];
+        tvm::Tensor computed_tensor = out_info[i];
+        // TODO(srkreddy1238): To be verified before enabling it.
+        // Especially for operator like cast with different dtype for input and output.
+        // CHECK_EQ(inferred_tensor->dtype, computed_tensor->dtype);
+        for (uint32_t j = 0; j < inferred_tensor->shape.size(); j++) {
```
Comment (tqchen) [inline src/compiler/compile_engine.cc]: prefer ++j

Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: oh, missed it :). Done now.

File: src/compiler/compile_engine.cc (hunk 4 above)
Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: Done

File: src/compiler/compile_engine.cc (hunk 3 above)
Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: I don't understand here. No need to check shapes ??

Comment (tqchen) [reply src/compiler/compile_engine.cc]: I mean the comment, no need to check data types equal

Comment (tqchen) [reply src/compiler/compile_engine.cc]: I am just confusing about the TODO left here.

Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: oh, ok

Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: Done

File: src/compiler/compile_engine.cc [hunk 6]
```diff
@@ -202,6 +202,22 @@ class CompileEngine {
       Array<Tensor> out = fcompute[inode.source->op()](
           inode.source->attrs, op_inputs, out_info);
       CHECK_EQ(out.size(), inode.source->num_outputs());
+
+      // check output dimentions also match
+      // This check is to make sure the NNVM operator Infer match with Compute result.
+      // Missing this check may pass the build but leads to runtime errors.
+      for (uint32_t i=0; i < out.size(); ++i) {
```
Comment (tqchen) [inline src/compiler/compile_engine.cc]: nit: space(i = 0)

Comment (srkreddy1238) [reply src/compiler/compile_engine.cc]: Done

**Pull request merged**